device_doctor/lib/src/android_device.dart

```diff
@@ -215,4 +222,40 @@ class AndroidDevice implements Device {
   Future<void> recover() async {
     await eval('adb', <String>['-s', deviceId, 'reboot'], canFail: false);
   }
+
+  @override
+  Future<void> prepare() async {
+    await killProcesses();
+  }
+
+  /// Kill top running process if existing.
+  @visibleForTesting
+  Future<bool> killProcesses({ProcessManager processManager}) async {
+    processManager ??= LocalProcessManager();
+    String result;
+    result = await eval('adb', <String>['shell', 'dumpsys', 'activity', '|', 'grep', 'top-activity'],
+        canFail: true, processManager: processManager);
+
+    // Skip uninstalling process when no device is available or no application exists.
+    if (result == 'adb: no devices/emulators found' || result == null || result.isEmpty) {
+      stdout.write('no process is running');
+      return true;
+    }
+    final List<String> results = result.trim().split('\n');
+    // Example result:
+    //
+    // Proc # 0: fore  T/A/T  trm: 0 4544:com.google.android.googlequicksearchbox/u0a66 (top-activity)
+    final RegExp processRegExp = RegExp(' 0 [0-9]+:(?<process>.+)/');
```

Comment:
Will that `0` always be there?  I'm not sure what it represents.  `trm: \d+ \d+:(.+)/` would be more robust.

Reply:
Was following the [first answer](https://stackoverflow.com/questions/31079853/how-do-i-kill-all-active-tasks-apps-using-adb/31079988). 
Now switched to use index instead of regexp to extract the process more accurate.